cub/cub/device/dispatch/dispatch_reduce.cuh

```diff
@@ -758,34 +761,51 @@ struct DispatchSegmentedReduce
         break;
       }
 
+      const auto num_segments_per_invocation =
+        static_cast<::cuda::std::int64_t>(::cuda::std::numeric_limits<std::int32_t>::max());
+      const ::cuda::std::int64_t num_invocations = ::cuda::ceil_div(num_segments, num_segments_per_invocation);
+      for (::cuda::std::int64_t invocation_index = 0; invocation_index < num_invocations; invocation_index++)
+      {
+        const auto current_seg_offset = invocation_index * num_segments_per_invocation;
+        const auto num_current_segments =
+          ::cuda::std::min(num_segments_per_invocation, num_segments - current_seg_offset);
+
+        auto current_begin_offset = detail::offset_input_iterator{
+          d_begin_offsets, THRUST_NS_QUALIFIER::constant_iterator<const ::cuda::std::int64_t>{current_seg_offset}};
```

Comment:
Q: Why can't we simply move the iterator here?
```suggestion
        auto current_begin_offset = d_begin_offsets + current_seg_offset;
```

Reply:
Good question! The reason is that `advance` may  not be implemented as a host function.

Reply:
I have the feeling you are not talking about `cuda::std::advance`. Which `advance` are you talking about?

I also have the feeling that this function should work on host and device.

Reply:
Is this a realistic use case we support? The ability to advance an iterator sounds like an absolute requirement to me.

Reply:
Let's leave the workaround in this PR for now.